File: src/java_impl/TimeFormatter.java
package java_impl;

/**
 * Class provides methods to format a given time period of milliseconds
 * into a better-looking and more user-friendly string.
 */
public class TimeFormatter {

    /**
     * Static object reference to the main instance that is returned by using the getInstance method.
     * Made private to enable the Singleton design pattern.
     */
    private static TimeFormatter instance;

    /**
     * @return instance of the TimeFormatter class
     */
    public static TimeFormatter getInstance()
    {
        if(instance == null)
            instance = new TimeFormatter();

        return instance;
    }

    /**
     * Primary constructor made private to enable the Singleton design pattern.
     */
    private TimeFormatter(){}

    /**
     * By default, this method will format the provided time period by including all the time units.
     *
     * @param millis time period to be formatted
     * @param timeFormat formatting type to be used while formatting the string
     *
     * @return formatted string of the provided time period
     */
    public String format(long millis, TimeFormat timeFormat)
    {
        return timeFormat.format(millis, 99);
    }

    /**
     * Method can limit the number of time units that appear in the final string, by counting down from the largest time unit.
     *
     * @param millis time period to be formatted
     * @param timeFormat formatting type to be used while formatting the string
     * @param significantUnits number of time units to be shown in the formatted string
     *
     * @return formatted string of the provided time period
     */
    public String format(long millis, TimeFormat timeFormat, int significantUnits)
    {
        return timeFormat.format(millis, significantUnits);
    }
}

File: src/kotlin_impl/TimeFormatter.kt
package kotlin_impl

/**
 * Class provides methods for formatting time periods
 *
 * Several formats available:
 *
 * 1) Verbal
 * Ex. 102 days, 22 hours, 33 minutes and 21 seconds
 *
 * 2) Single suffix
 * Ex. 22h:33m:21s
 *
 ************************
 * More to be implemented
 */
object TimeFormatter {

    /**
     * By default, this method will format the given time duration and will include all possible time units
     *
     * @param millis time duration given in milliseconds
     * @param format the format type according to which the time duration is to be formatted
     *
     * @return formatted string of the given duration in milliseconds
     */
    fun format(millis:Long, format : TimeFormat) : String
    {
        return format.format(millis, 99)
    }

    /**
     * @param millis time duration given in milliseconds
     * @param format the format type according to which the time duration is to be formatted
     * @param significantUnits the number of units to be included in the string
     *
     * @return formatted string of the given duration in milliseconds
     */
    fun format(millis:Long, format: TimeFormat, significantUnits:Int):String{

        return format.format(millis, significantUnits)
    }
}